src/handlers/push_message.rs

```diff
@@ -54,13 +55,92 @@ pub async fn handler(
     headers: HeaderMap,
     RequireValidSignature(Json(body)): RequireValidSignature<Json<PushMessageBody>>,
 ) -> Result<Response> {
-    let request_id = get_req_id(&headers);
+    let (res, analytics_option) = handler_internal(
+        Path((tenant_id, id)),
+        StateExtractor(state),
+        headers,
+        RequireValidSignature(Json(body)),
+    )
+    .await;
 
-    increment_counter!(state.metrics, received_notifications);
+    let (status, response) = match res {
+        Ok(res) => (res.status_code.clone().as_u16(), res),
+        Err(error) => {
+            let res = error.into_response();
+            let status_code = res.status().clone().as_u16();
+            (status_code, res)
+        }
+    };
+
+    #[cfg(feature = "analytics")]
+    if let Some(mut message_info) = analytics_option {
+        message_info.status = status;
+        message_info.success = status >= 200 && status < 300;
+        message_info.response_message = Some(response.body().to_string());
+
+        tokio::spawn(async move {
+            if let Some(analytics) = &state.analytics {
+                let (country, continent, region) = analytics
+                    .geoip
+                    .lookup_geo_data(addr.ip())
+                    .map_or((None, None, None), |geo| {
+                        (geo.country, geo.continent, geo.region)
+                    });
+
+                debug!(
+                    %request_id,
+                    %tenant_id,
+                    client_id = %id,
+                    ip = %addr.ip(),
+                    "loaded geo data"
+                );
+
+                message_info.country = country;
+                message_info.continent = continent;
+                message_info.region = region.map(|r| Arc::from(r.join(", ")));
+
+                analytics.message(message_info);
+            }
+        });
+    }
+
+    Ok(response)
+}
+
+pub async fn handler_internal(
+    Path((tenant_id, id)): Path<(String, String)>,
+    StateExtractor(state): StateExtractor<Arc<AppState>>,
+    headers: HeaderMap,
+    RequireValidSignature(Json(body)): RequireValidSignature<Json<PushMessageBody>>,
+) -> (Result<Response>, Option<MessageInfo>) {
+    #[cfg(feature = "analytics")]
+    let topic: Option<Arc<str>> = body.payload.topic.as_ref().map(|t| t.clone().into());
 
     #[cfg(feature = "analytics")]
     let (flags, encrypted) = (body.payload.flags, body.payload.is_encrypted());
 
+    #[cfg(feature = "analytics")]
+    let mut analytics = MessageInfo {
+        msg_id: body.id.into(),
+        region: None,
+        country: None,
+        continent: None,
+        project_id: tenant_id.into(),
+        client_id: id.into(),
+        topic,
+        push_provider: client.push_type.as_str().into(),
+        encrypted,
+        flags,
+        status: 0,
+        success: false,
```

Comment:
One more reason to remove this field. Someone someday will just do a count on 'success=false' and panic because there are a lot of failures :)

Reply:
done